_posts/2025-02-28-lattice-bitset.md

```diff
@@ -0,0 +1,801 @@
+    return (left & right) == left && spec_is_subtype(left.spec, right.spec);
+}
+```
+
+If we decompose the problem that way, we can write some lattice operations for
+`Spec`. Let's start with `is_subtype`:
+
+```c++
+// Ask if `left` is a subtype of `right`
+bool spec_is_subtype(struct Spec left, struct Spec right) {
+    if (right.spec_kind == SpecTop || left.spec_kind == SpecBottom) {
+        // Top is a supertype of everything and Bottom is a subtype of
+        // everything
+        return true;
+    }
+    // left is now either SpecInt or SpecTop
+    // right is now either SpecInt or SpecBottom
+    // The only way left could be a subtype of right is if they are both
+    // SpecInt and the value is the same
+    if (left.spec_kind == SpecInt &&
+        right.spec_kind == SpecInt &&
+        left.value == right.value) {
+        return true;
+    }
+    return false;
+}
+```
+
+That takes a bit of time to internalize, so please read over it a couple of
+times and work out the cases by hand. It's really useful for implementing
+`spec_join`:
+
+```c
+struct Spec spec_join(struct Spec left, struct Spec right) {
+    if (spec_is_subtype(left, right)) { return right; }
+    if (spec_is_subtype(right, left)) { return left; }
+    // We know that neither left nor right is either SpecTop or SpecBottom
+    // because that would have been covered in one of the subtype cases, so
+    // we're join-ing two SpecInts. That's SpecTop.
+    struct Spec result;
+    result.spec_kind = SpecTop;
+    return result;
+}
+```
+
+It's a couple more instructions than a single bitwise operation and a compare,
+but it's still compact and fast.
+
+Let's talk about some problems you might run into while using this API.
+
+## Bottom API
+
+A common mistake when handling `Bottom` is treating it as an error, or assuming
+it will never show up in your program. To expand on its brief introduction
+above, the two main consequences of `Bottom`'s place at the bottom of the type
+lattice are:
+
+1. A value of type `Bottom` can never be *defined*. An instruction with an
+   output of type `Bottom` will therefore never define its output. Such an
+   instruction might loop infinitely, it might crash your program, or it might
+   jump to another location in the program (if your compiler supports
+   control-flow instructions that define values).
+2. A value of type `Bottom` can never be *used*, so an instruction with an
+   input of type `Bottom` is unreachable. This follows from the previous item
+   (since you obviously can't use a value you can't define), but is worth
+   calling out separately.
+
+These two properties make checking for `Bottom` useful in an unreachable code
+elimination pass. Otherwise, though, `Bottom` is just another type, and your
+type predicates will generally be cleaner if you design them to correctly
+handle `Bottom` without explicitly testing for it.
+
+Most of the time, `Bottom` support happens out naturally:
+
+```c++
+if (t <= TLongExact) {
```

Comment:
Is this an overloaded operator? Shouldn't it be an `is_subtype()` call? Also, I need some introduction to what TLongExact is, and specifically what the conceptual difference between a subtype and strict subtype of it.